Add docs on the recreate procedure (#1819)
neo4j/docs-operations: modules/ROOT/pages/clustering/databases.adoc

+----------------------------------------------------------------------------------------------------------------------------------------+
----

[role=label--new-5.24]
[[recreate-databases]]

> == Recreate databases using a procedure

Reviewer: I don't think we need to specify this is a procedure here, since this is not done for Deallocate databases or Reallocate databases even though both of those are procedures as well.

> == Recreate databases using a procedure

Author: ```suggestion
== Recreate databases
```

Neo4j 5.24 introduces the xref:reference/procedures.adoc#procedure_dbms_cluster_recreateDatabase[`dbms.cluster.recreateDatabase()`] procedure, which allows you:

* To restore a database store to a specified backup, while keeping all the users and roles (metadata) for the database.

> * To make a database writable again after it has been lost due to a disaster.

Author: ```suggestion
* To restore a database to a write mode again after it has been lost due to a disaster.
```

> * To restore a database to a write mode again after it has been lost due to a disaster.

Author: ```suggestion
* To restore a database to a write mode again after it has been lost (for example, due to a disaster).
```

> * To restore a database to a write mode again after it has been lost (for example, due to a disaster).

Reviewer: On both this and the previous one we have the word "restore". This is unfortunately very overloaded, could we find a nice way of saying it without this word?

> See xref:clustering/disaster-recovery.adoc[] for more information.

Author: ```suggestion
// See xref:clustering/disaster-recovery.adoc[] for more information.
```

> See xref:clustering/disaster-recovery.adoc[] for more information.

Reviewer: I think we should hold off on pointing to the disaster recovery docs here, until we have made the changes there.

 
[CAUTION]
====
The recreate procedure works only for real user databases and not for composite databases, or the `system` database.

Keep in mind the recreate procedure results in downtime while the stores get updated.
The time is unbounded and may depend on the store size and network speed.

> The recreate procedure selects the most up-to-date seed from the defined seeding servers.

Reviewer: Do we think this is a bit to advanced for this documentation, and that the spirit of what will happen is still here with this sentence?

> The recreate procedure selects the most up-to-date seed from the defined seeding servers.

Reviewer: Natalia will follow up on this.

If only a subset of the previous cluster members is specified as seeders, there is a risk of data loss.

> To avoid this, ensure you back up the current store before replacing it with a seed from a URI, so the latest state can be restored if needed.

Reviewer: This is not really related to the sentence above. There is a risk of data loss when recreating the database if you choose you seeders "wrong" or the system database is not up to date. However, this sentence is more related to the first use case of recreate, when you want to go back in time.

====

The database in question can be in an `online` or `offline` state when it is recreated, but a successful operation starts the database regardless of the previous state.

If CDC is enabled for the database, it will still be enabled after recreating the database, however the chain will be broken.

> See link:https://neo4j.com/docs/cdc/current/[CDC documentation] for more details.

Author: This is a general link to the docs. Do we have any more specific guidelines on the topic?

Reviewer: Not that I know of, but I can check that out

Reviewer: Yes, I think here: [CDC](https://development.neo4j.dev/docs/cdc/current/backup-restore/) however, might need some additional info to make it a smooth experience.

Reviewer: I'll look into adding some more details

> See link:https://neo4j.com/docs/cdc/current/[CDC documentation] for more details. - general link ?? Do we have any more specific guidelines on the topic?

Author: ```suggestion
See link:https://neo4j.com/docs/cdc/current/[CDC documentation] for more details.
```

Author: This is a general link to the docs. Do we have any more specific guidelines on the topic?

Before recreating a database, any eventual quarantined states need to be addressed.

> For more information, refer to the quarantined state documentation.

Author: I didn't find where to link this place.

> For more information, refer to the quarantined state documentation.

Author: ```suggestion
For more information, see xref:clustering/databases.adoc#recreate-databases[Standard databases -> Error handling].
```

> ?? For more information, refer to the Quarantined State documentation. ??

Author: ```suggestion
For more information, refer to the quarantined state documentation.
```

Author: I didn't find where to link this paragraph.

You need xref:authentication-authorization/dbms-administration.adoc#access-control-dbms-administration-database-management[the DBMS `DATABASE MANAGEMENT` privileges] to run the recreate procedure. 

[[recreate-seeding-options]]
=== Seeding options and configurations

> The store to be used during the recreation can be defined in three different ways.

Author: ```suggestion
The database store to be used during the recreation can be defined in three different ways.
```

Reviewer: ```suggestion
The store to be used during the recreation of a database can be defined in three different ways.
```

One method uses a backup, while the two others use available allocations in the cluster.

Additionally, you have the option to modify <<alter-topology-recreate, the topology>> during the recreation process.
However, note that the store format, access, and enrichment cannot be altered during recreation.

[[uri-seed]]
==== `seedURI`

If you provide a URI to a backup or a dump, the stores on all allocations will be replaced by the backup or the dump at the given URI.
The new allocations can be put on any `ENABLED` server in the cluster.
See <<cluster-seed-uri, Seed from URI>> for more details.

[source, shell]
----

> CALL dbms.cluster.recreateDatabase("neo4j", {seedURI: "myBackup"});

Reviewer: I know this was my suggestion in the google doc, but how does this conform with other examples in the documentation? Is it really ok to write "myBackup" for example? Also, another example I saw used the db name "foo", is this also something we should conform to here?

Reviewer: Made it clearer so should be ok

----

[[seed-servers]]
==== `seedingServers`

Specify a set of available servers.
The stores on all allocations will be synchronized to the most up-to-date store from the defined servers.
The number of defined servers cannot exceed the number of total allocations in the desired topology. 

[source, shell]
----
CALL dbms.cluster.recreateDatabase("neo4j", {seedingServers: ["serverId1", "serverId2", "serverId3"]});
----

[[undefined-servers]]
==== Undefined servers

> If you provide an empty list of seeding servers or do not specify a `seedURI`, Neo4j automatically selects all available allocations of the database as seeders.

Author: ```suggestion
If you provide an empty list of seeding servers and do not specify a `seedURI`, Neo4j automatically selects all available allocations of the database as seeders.
```

Reviewer 2: and

Either `seedURI` or `seedingServers` must be in the options. If you define neither that is an error.
If you define both then:
* either `seedingServers` is an empty list then see "Undefined servers with fallback backup"
* or `seedingServers` is not empty, that is an error

Reviewer: The errors are being put in further up.

+----------------------------------------------------------------------------------------------------------------------------------------+
----

[role=label--new-5.24]
[[recreate-databases]]
== Recreate databases

Neo4j 5.24 introduces the xref:reference/procedures.adoc#procedure_dbms_cluster_recreateDatabase[`dbms.cluster.recreateDatabase()`] procedure, which allows you to restore a database:

* to a specified backup, while keeping all the associated privileges for the database.

* to a write mode again after it has been lost (for example, due to a disaster).
// See xref:clustering/disaster-recovery.adoc[] for more information.
 
[CAUTION]
====
The recreate procedure works only for real user databases and not for composite databases, or the `system` database.

Keep in mind the recreate procedure results in downtime while the stores get updated.
The time is unbounded and may depend on different factors -- for example, the size of the store, network speed, etc.

The database recreation process is designed to select the latest seed from the defined seeding servers.

> Specifying only a subset of the previous cluster servers as seeders carries a risk of data loss.

Reviewer: Do we need to think about the dataloss part a bit more.

----

[role=label--new-5.24]
[[recreate-databases]]
== Recreate a database

Neo4j 5.24 introduces the xref:reference/procedures.adoc#procedure_dbms_cluster_recreateDatabase[`dbms.cluster.recreateDatabase()`] procedure, which allows you:

* To change the database store to a specified backup, while keeping all the associated privileges for the database.

* To make your database write-available again after it has been lost (for example, due to a disaster).
// See xref:clustering/disaster-recovery.adoc[] for more information.
 
[CAUTION]
====
The recreate procedure works only for real user databases and not for composite databases, or the `system` database.

Remember that the recreate procedure results in downtime while the stores get updated.
The time is unbounded and may depend on different factors -- for example, the size of the store, network speed, etc.
====

The database in question can be in an `online` or `offline` state when it is recreated, but a successful operation starts the database regardless of its previous state.

If CDC is enabled for the database, it will still be enabled after recreating the database, however the chain will be broken.

> See link:https://neo4j.com/docs/cdc/current/backup-restore/[CDC documentation -> Restore backups and snapshots] for more details.

Reviewer: Okay, so I think we should not link to the docs I suggested last time. That handles what the user can see and should do if the database was restored. In the case of recreate, we change database id , so effectively it is a new database. Therefore we should link to how to set CDC up from the beginning. My suggestion is this: 

If CDC is enabled for the database, the chain will stop when the database is recreated. While the recreated database will still have CDC enabled, follow the link:https://development.neo4j.dev/docs/cdc/current/existing-databases/[Initialize CDC applications from an existing database] guide to get it working again.

Before recreating a database, any eventual quarantined states need to be addressed.
For more information, see xref:database-administration/standard-databases/errors.adoc#quarantine[Standard databases -> Error handling].

You need xref:authentication-authorization/dbms-administration.adoc#access-control-dbms-administration-database-management[the `CREATE DATABASE` and `DROP DATABASE` privileges] to run the recreate procedure.

To check if the recreation is successful, use the `SHOW DATABASES` command and verify that all allocations have been started.

Additionally, you have the option to modify <<alter-topology-recreate, the topology>> during the recreation process.
However, note that the store format, access, and enrichment cannot be altered during recreation.

[[recreate-seeding-options]]
=== Seeding options

The store to be used during the recreation of a database can be defined in different ways.
One method uses a backup, while others use available allocations in the cluster.

You can use either <<uri-seed, `seedURI`>> or <<seed-servers, `seedingServers`>> to specify the source from which the database should be recreated.

* If you define neither, an error is thrown.

> * If you define both of them, then `seedingServers` must be an empty list.

Author: ```suggestion
* If you define both of them, then `seedingServers` must be an empty list.
See <<undefined-servers-backup, Undefined servers with fallback backup>> for more details.
```

* If `seedingServers` is not empty and `seedURI` is also defined, an error will occur.

> See <<undefined-servers-backup, Undefined servers with fallback backup>> for more details.

Author: ```suggestion
```

[[uri-seed]]
==== Use backup as a seed

If you provide a URI to a backup or a dump, the stores on all allocations will be replaced by the backup or the dump at the given URI.
The new allocations can be put on any `ENABLED` server in the cluster.
See <<cluster-seed-uri, Seed from URI>> for more details.

> To avoid possible data loss, ensure you back up the current store before replacing it with a seed from a URI, so the latest state can be restored if needed.

Author: ```suggestion
```

[source, shell]
----
CALL dbms.cluster.recreateDatabase("neo4j", {seedURI: "s3:/myBucket/myBackup.backup"});
----

[[seed-servers]]
==== Use available allocations as seed

The database recreation process is designed to select the latest seed from the seeding servers.

> Specifying only a subset of the previous cluster servers as seeders carries a risk of data loss.

Reviewer: What about this suggestions:

After the recreation has finished, the database will have the most up-to-date store that could be found on the seeding servers. 

And after that in a "caution" block:
Recreation is based on remaining stores or store defined by the user. This means that stores which were lost or not defined is not used for the recreation. If not used stores were more up to date than the used ones, this will result in data loss.

The new allocations can be put on any `ENABLED` server in the cluster.
See <<cluster-seed-uri, Seed from URI>> for more details.

[source, shell]
----
CALL dbms.cluster.recreateDatabase("neo4j", {seedURI: "s3:/myBucket/myBackup.backup"});
----

[[seed-servers]]
==== Use available servers as a seed

After the recreation is complete, the database will have the latest data store from the seeding servers.

[CAUTION]
====
Recreation is based on remaining stores or a store defined by the user.
This means that stores which were lost or not defined are not used for the recreation.
If not used stores were more up to date than the used ones, this results in data loss.
====

[[specified-servers]]
===== Specified servers

You can specify a set of available servers.
The stores on all allocations will be synchronized to the most up-to-date store from the defined servers.
The number of defined servers cannot exceed the number of total allocations in the desired topology. 

[source, shell]
----
CALL dbms.cluster.recreateDatabase("neo4j", {seedingServers: ["serverId1", "serverId2", "serverId3"]});
----

[[undefined-servers]]
===== Undefined servers

> If you provide an empty list of seeding servers and do not specify a `seedURI`, Neo4j automatically selects all available allocations of the database as seeders.

Reviewer: Here we use "Neo4j automatically selects" and down below we use "the system finds" I think we should stick with one of them and use it in both places.

The store will be replaced by the most up-to-date seeder available in the cluster.

[source, shell]
----
CALL dbms.cluster.recreateDatabase("neo4j", {seedingServers: []});
----

[[undefined-servers-backup]]
===== Undefined servers with fallback backup

> If both an empty list of seeding servers and a `seedURI` are provided, the system finds all available allocations of the database and use those as seeders.

Author: ```suggestion
If both an empty list of seeding servers and a `seedURI` are provided, Neo4j finds all available allocations of the database and use those as seeders.
```
